### PR TITLE
 STORM-2438: added in rebalance changes to support RAS

### PR DESCRIPTION
**Pull request by revans2**
This is based off #2339 (which is in the first commit).  If people want to review it all together I am fine with closing the other one.

This does a few things

1) It updates the AsyncLocalizer and Slot so that when the jar, conf, or serialized topology changes the new code is downloaded to a temp location, the workers are shot, the changes are made, and the workers are brought back up.
2) It adds in changes to the rebalance command so you can adjust config options and resources in addition to what was there before.

I plan on adding support for generic blobs to optionally cause the workers to be restarted on update. This would be for things like jars on the classpath.  I plan to do this work in a separate pull request because this is already very large.

### CODE REVIEW

**Comment by revans2** [comment]
I have done a lot of manual testing with this patch, but if others want to try it out feel free to kick the tires.  Any feedback is appreciated.

**Comment by revans2** [comment]
For anyone interested I just rebased as #2339 was merged into master.

**Comment by revans2** [comment]
For anyone interested I just rebased to fix a very minor merge conflict.

File: storm-server/src/main/java/org/apache/storm/daemon/supervisor/Slot.java [hunk 1]
```diff
@@ -104,9 +111,12 @@
             this.port = port;
             this.iSupervisor = iSupervisor;
             this.localState = localState;
+            this.changingCallback = changingCallback;
         }
     }
-    
+
+    //TODO go through all of the state transitions and make sure we handle changingBlobs
+    //TODO make sure to add in transition helpers that clean changingBlobs && pendingChangeingBlobs for not the current topology
```
Comment (kishorvpatil) [inline storm-server/src/main/java/org/apache/storm/daemon/supervisor/Slot.java]: ??

File: storm-server/src/main/java/org/apache/storm/daemon/supervisor/Slot.java [hunk 2]
```diff
@@ -263,7 +314,41 @@ public String toString() {
             return "{ " + topoId + ": " + request + " }";
         }
     }
-    
+
+    /**
+     * Holds the information about a blob that is changing.
+     */
+    static class BlobChangeing {
```
Comment (kishorvpatil) [inline storm-server/src/main/java/org/apache/storm/daemon/supervisor/Slot.java]: Does this need to be `BlobChanging`?

File: storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java [hunk 3]
```diff
@@ -156,17 +146,141 @@ public AsyncLocalizer(Map<String, Object> conf, AtomicReference<Map<Long, LocalA
...
+                            failures++;
+                            if (failures > blobDownloadRetries) {
+                                throw new RuntimeException("Could not download...", e);
+                            }
+                            LOG.warn("Failed to download blob {} will try again in {} ms", blob, ATTEMPTS_INTERVAL_TIME, e);
+                            Utils.sleep(ATTEMPTS_INTERVAL_TIME);
+                        }
+                    }
+                }
+                LOG.debug("FINISHED download of {}", blob);
```
Comment (kishorvpatil) [inline storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java]: Should this be info?

File: storm-server/src/main/java/org/apache/storm/localizer/LocallyCachedBlob.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+                //TODO we probably want to not use this, or make it just return something that has less power to modify things
+                cb.blobChanging(pna.getAssignment(), pna.getPort(), this, gtg);
+            } finally {
+                gtg.countDownIfLatchWasNotGotten();
+            }
+        }
+        try {
+            cdl.await(3, TimeUnit.MINUTES);
+        } catch (InterruptedException e) {
+            //TODO need to think about error handling here in general.
```
Comment (kishorvpatil) [inline storm-server/src/main/java/org/apache/storm/localizer/LocallyCachedBlob.java]: Should we at least log the error being ignored?

File: storm-server/src/main/java/org/apache/storm/daemon/supervisor/Slot.java (hunk 1 above)
Comment (revans2) [reply storm-server/src/main/java/org/apache/storm/daemon/supervisor/Slot.java]: Yes I left in a lot of TODOs #2363 cleans them up, but I will try to pull back what I can here.

In this particular case I did go through the transitions and I think I have it all covered.

File: storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java (hunk 3 above)
Comment (revans2) [reply storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java]: I used it for debugging.  If you see a good reason for it to be info I can make it info, but I didn't see a lot of value in it.

**Comment by revans2** [comment]
@kishorvpatil I think I addressed your review comments.  If you really want the log message to be info I can do it.

**Comment by kishorvpatil** [comment]
@revans2 LGTM.

**Pull request merged**